pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,251 @@
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	lock              sync.Mutex
+	kvmController     *device_manager.DeviceController
+	configMapInformer cache.SharedIndexInformer
+	nodeInformer      cache.SharedIndexInformer
+	clientset         kubecli.KubevirtClient
+	Queue             workqueue.RateLimitingInterface
+	host              string
+	namespace         string
+	logger            *log.FilteredLogger
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, nodeInformer, configMapInformer cache.SharedIndexInformer, clientset kubecli.KubevirtClient, host, namespace string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController:     kvmController,
+		configMapInformer: configMapInformer,
+		nodeInformer:      nodeInformer,
+		clientset:         clientset,
+		Queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		host:      host,
+		namespace: namespace,
+		logger:    log.DefaultLogger(),
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
+	defer n.Queue.ShutDown()
+	n.logger.Infof("node-labeller is running")
+
+	if !n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+		n.logger.Infof("Node-labeller cannot work without KVM device.")
+		return
+	}
+
+	n.configMapInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
+		},
+		DeleteFunc: func(obj interface{}) {
+			n.Queue.AddRateLimited(obj)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			n.Queue.AddRateLimited(new)
+		},
+	})
+
+	go n.configMapInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.configMapInformer.HasSynced)
+
+	n.nodeInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(old, new interface{}) {
+			n.Queue.AddRateLimited(new)
+		},
+	})
+
+	go n.nodeInformer.Run(stop)
+	cache.WaitForCacheSync(stop, n.nodeInformer.HasSynced)
+
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(n.runWorker, time.Second, stop)
+	}
+	<-stop
+}
+func (n *NodeLabeller) runWorker() {
+	for n.Execute() {
+	}
+}
+
+func (n *NodeLabeller) Execute() bool {
+	key, quit := n.Queue.Get()
+	if quit {
+		return false
+	}
+	defer n.Queue.Done(key)
+
+	var (
+		node *v1.Node
+		ok   bool
+	)
+
+	//If event comes from Node, check if the node is the same as virt-handler is running on.
```

Comment:
can we make the informer only listen for updates for the specific node this component is running on? That way we'll only receive updates for a single node.

a field selector like this maybe?

```fieldSelector := fields.OneTermEqualSelector("metadata.name", "whatever the local node name is")```

and use that for the node informer.